### ISSUE TITLE
DistributedDataParallel throws RuntimeError for sparse embeddings

### ISSUE THREAD

**Issue by mrshenli** [pytorch/pytorch#17356]
## 🐛 Bug

Using `nn.Embedding(n, d, sparse=True)` in `DistributedDataParallel` models hits the following error in the backward pass.

CC: @glample  @soumith 

```
Traceback (most recent call last):
  File "/home/shenli/local/miniconda/lib/python3.7/multiprocessing/process.py", line 297, in _bootstrap
    self.run()
  File "/home/shenli/local/miniconda/lib/python3.7/multiprocessing/process.py", line 99, in run
    self._target(*self._args, **self._kwargs)
  File "dist_bug.py", line 32, in init_processes
    y.mean().backward()
  File "/home/shenli/project/pytorch/torch/tensor.py", line 106, in backward
    torch.autograd.backward(self, gradient, retain_graph, create_graph)
  File "/home/shenli/project/pytorch/torch/autograd/__init__.py", line 93, in backward
    allow_unreachable=True)  # allow_unreachable flag
  File "/home/shenli/project/pytorch/torch/nn/parallel/distributed.py", line 450, in distributed_data_parallel_hook
    self._queue_reduction(bucket_idx)
  File "/home/shenli/project/pytorch/torch/nn/parallel/distributed.py", line 480, in _queue_reduction
    self.device_ids)
RuntimeError: sparse tensors do not have is_contiguous (is_contiguous at ../aten/src/ATen/SparseTensorImpl.cpp:47)
frame #0: c10::Error::Error(c10::SourceLocation, std::string const&) + 0x45 (0x7f1225ae6b65 in /home/shenli/project/pytorch/torch/lib/libc10.so)
frame #1: at::SparseTensorImpl::is_contiguous() const + 0x7d (0x7f122629262d in /home/shenli/project/pytorch/torch/lib/libcaffe2.so)
frame #2: at::native::contiguous(at::Tensor const&) + 0x20 (0x7f1226434310 in /home/shenli/project/pytorch/torch/lib/libcaffe2.so)
frame #3: at::TypeDefault::contiguous(at::Tensor const&) const + 0x50 (0x7f122667cf80 in /home/shenli/project/pytorch/torch/lib/libcaffe2.so)
frame #4: torch::autograd::VariableType::contiguous(at::Tensor const&) const + 0x10f (0x7f12238c9e0f in /home/shenli/project/pytorch/torch/lib/libtorch.so.1)
frame #5: <unknown function> + 0x4a4fe8 (0x7f123bb9ffe8 in /home/shenli/project/pytorch/torch/lib/libtorch_python.so)
frame #6: c10d::queueReduction(c10d::ProcessGroup&, std::vector<std::vector<at::Tensor, std::allocator<at::Tensor> >, std::allocator<std::vector<at::Tensor, std::allocator<at::Tensor> > > >&, std::vector<long, std::allocator<long> > const&) + 0x44e (0x7f123bd0f73e in /home/shenli/project/pytorch/torch/lib/libtorch_python.so)
frame #7: <unknown function> + 0x61272b (0x7f123bd0d72b in /home/shenli/project/pytorch/torch/lib/libtorch_python.so)
frame #8: <unknown function> + 0x11a513 (0x7f123b815513 in /home/shenli/project/pytorch/torch/lib/libtorch_python.so)
<omitting python frames>
frame #18: torch::autograd::PyFunctionPostHook::operator()(std::vector<torch::autograd::Variable, std::allocator<torch::autograd::Variable> > const&, std::vector<torch::autograd::Variable, std::allocator<torch::autograd::Variable> > const&) + 0xc8 (0x7f123ba45d28 in /home/shenli/project/pytorch/torch/lib/libtorch_python.so)
frame #19: torch::autograd::Engine::evaluate_function(torch::autograd::FunctionTask&) + 0x1574 (0x7f12233beea4 in /home/shenli/project/pytorch/torch/lib/libtorch.so.1)
frame #20: torch::autograd::Engine::thread_main(torch::autograd::GraphTask*) + 0xc0 (0x7f12233bfab0 in /home/shenli/project/pytorch/torch/lib/libtorch.so.1)
frame #21: torch::autograd::Engine::thread_init(int) + 0x81 (0x7f12233bcaa1 in /home/shenli/project/pytorch/torch/lib/libtorch.so.1)
frame #22: torch::autograd::python::PythonEngine::thread_init(int) + 0x2a (0x7f123ba335ea in /home/shenli/project/pytorch/torch/lib/libtorch_python.so)
frame #23: <unknown function> + 0xb8678 (0x7f123c6ad678 in /home/shenli/local/miniconda/lib/libstdc++.so.6)
frame #24: <unknown function> + 0x7e25 (0x7f1253a59e25 in /lib64/libpthread.so.0)
frame #25: clone + 0x6d (0x7f1253783bad in /lib64/libc.so.6)

```

## To Reproduce

```python
import os
import torch
import torch.nn as nn
from torch.multiprocessing import Process

class Model(nn.Module):
    def __init__(self):
        super(Model, self).__init__()
        self.embedding = nn.Embedding(10, 3, sparse=True)
        self.net = nn.Linear(2, 3)

    def forward(self, x):
        return self.net(x) + self.embedding(torch.tensor(0).cuda())


def init_processes(rank, size, backend='gloo'):
    os.environ['MASTER_ADDR'] = '127.0.0.1'
    os.environ['MASTER_PORT'] = '29500'
    torch.distributed.init_process_group(backend, rank=rank, world_size=size)
    torch.cuda.set_device(rank)
    model = nn.parallel.DistributedDataParallel(Model().cuda(), device_ids=[rank], output_device=0)
    x = torch.rand(20, 2).cuda()
    y = model(x)
    y.mean().backward()


if __name__ == "__main__":
    size = 2
    processes = []
    for rank in range(size):
        p = Process(target=init_processes, args=(rank, size))
        p.start()
        processes.append(p)

    for p in processes:
        p.join()
```

## Environment

PyTorch version: 1.1.0a0+63214b5
Is debug build: No
CUDA used to build PyTorch: 9.2.88

OS: CentOS Linux 7 (Core)
GCC version: (GCC) 4.8.5 20150623 (Red Hat 4.8.5-28)
CMake version: version 3.12.2

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 9.2.88
GPU models and configuration: 
GPU 0: Tesla M40
GPU 1: Tesla M40

Nvidia driver version: 396.26
cuDNN version: Could not collect

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] torch==1.1.0a0+63214b5
[conda] blas                      1.0                         mkl  
[conda] mkl                       2019.1                      144  
[conda] mkl-include               2019.1                      144  
[conda] mkl_fft                   1.0.6            py37hd81dba3_0  
[conda] mkl_random                1.0.2            py37hd81dba3_0  
[conda] torch                     1.1.0a0+63214b5           dev_0    <develop>

**Comment by mrshenli** [comment]
The error occurs [here](https://github.com/pytorch/pytorch/blob/master/torch/csrc/distributed/c10d/ddp.cpp#L159), where we pass a sparse tensor to `flatten_dense_tensors` which tries to get a contiguous version of it. Ultimately, it  tries to run a  NCCL reduce operation on multiple sparse tensors using the `SUM` operation. @fmassa pointed out that it will not work, as a sparse tensor uses two dense tensors to store indices and values. To solve this problem, we need to implement a new reduce operation for sparse tensors. As it turns out to be a new feature instead of a bug, I will create a new issue for it.

**Comment by mrshenli** [comment]
Blocked by #17369

**Comment by pietern** [comment]
Will be addressed by #19443.
